modules/tinymce/src/plugins/directionality/main/ts/core/Direction.ts

```diff
@@ -5,28 +5,43 @@
  * For commercial licenses see https://www.tiny.cloud/
  */
 
-import Editor from 'tinymce/core/api/Editor';
-import Tools from 'tinymce/core/api/util/Tools';
+import { Arr } from '@ephox/katamari';
+import { Traverse, Attribute, SugarElement } from '@ephox/sugar';
 
-const setDir = (editor: Editor, dir: string) => {
-  const dom = editor.dom;
-  let curDir;
-  const blocks = editor.selection.getSelectedBlocks();
+import Editor from 'tinymce/core/api/Editor';
 
-  if (blocks.length) {
-    curDir = dom.getAttrib(blocks[0], 'dir');
+type Dir = 'rtl' | 'ltl';
 
-    Tools.each(blocks, (block) => {
-      // Add dir to block if the parent block doesn't already have that dir
-      if (!dom.getParent(block.parentNode, '*[dir="' + dir + '"]', dom.getRoot())) {
-        dom.setAttrib(block, 'dir', curDir !== dir ? dir : null);
+const setDir = (editor: Editor, dir: Dir) => {
+  const selectedBlocks = editor.selection.getSelectedBlocks();
+  Arr.each(selectedBlocks, (block) => {
+    const sugarBlock = SugarElement.fromDom(block);
+    const blockParent = Traverse.parent(sugarBlock);
+    blockParent.map((blockParent: SugarElement<Element>) => {
+      const blockParentDirection = Attribute.get(blockParent, 'dir');
+      if ((blockParentDirection === undefined) || (blockParentDirection !== dir)) {
+        setDirAttr(editor, sugarBlock, dir);
+      } else { // if parent and child dir are going to be the same then remove it from child
+        Attribute.remove(sugarBlock, 'dir');
+        editor.nodeChanged();
       }
     });
+  });
+};
 
-    editor.nodeChanged();
+const setDirAttr = (editor: Editor, element: SugarElement, dir: Dir): void => {
+  if (isListItem(element)) {
```

Comment:
This branch has a bug that will only be apparent with existing content. The problem is that it doesn't remove the directionality on the `li` elements. As such, if the `li` element already has a directionality it will never get changed and will be forever broken. As such we have 2 options here:
1. Ensure the `dir` attribute is removed on all child `li` elements of the list.
2. Remove the `dir` attribute for the element that's passed in. This is a little more brittle, as it will only remove the `dir` attribute from the `li` elements in the selection.

Also please make sure to add a test for trying to convert something like this (which is possible to do with the current version of TinyMCE):
```
<ul>
  <li dir="rtl">foo</li>
  <li>bar</li>
</ul>
```

Reply:
Yeah... I was thinking about the logic, we would have something like this, right?

```
each block
  if li
    li remove dir
    check parent of ul/ol (p2) that's not a li and has dir // dunno how we gonna do this
    if p2 dir = clickedDir
      p2 remove dir
    else
      set dir on parent of li 
  else
    if parent dir = clickedDir
      block remove dir
    else
      set dir on block
```

This line `check parent of ul/ol (p2) that's not a li and has dir` because we can have ul that has a li as a parent with the dir attribute (incorrectly):

```
<ul>
  <li dir="rtl">
    Item 1
    <ul> <!-- this one -->
      <li> 
        Item 2
      </li>
    </ul>
  </li>
</ul>
```

Reply:
I'm requesting a new review as I changed the logic to find the closest parent which has `dir` and is not a li.